wagtail/utils/utils.py

```diff
@@ -13,3 +13,22 @@ def deep_update(source, overrides):
         else:
             source[key] = overrides[key]
     return source
+
+
+def flatten_choices(choices):
+    """
+    Convert potentially grouped choices into a flat dict of choices.
+
+    flatten_choices([(1, '1st'), (2, '2nd')]) -> {1: '1st', 2: '2nd'}
+    flatten_choices([('Group', [(1, '1st'), (2, '2nd')])]) -> {1: '1st', 2: '2nd'}
+    """
+    ret = {}
+    for key, value in choices:
+        if isinstance(value, (list, tuple)):
+            # grouped choices (category, sub choices)
```

Comment:
Django's docs also show an example of using a dict to group the choices: https://docs.djangoproject.com/en/5.0/ref/models/fields/#field-choices

```
MEDIA_CHOICES = {
    "Audio": {
        "vinyl": "Vinyl",
        "cd": "CD",
    },
    "Video": {
        "vhs": "VHS Tape",
        "dvd": "DVD",
    },
    "unknown": "Unknown",
}
```

If we're handling the list/tuple case, I think it's reasonable to also handle the dict case 😄 

Also, as Dan said, this needs tests before we can merge it. We have the `EventPage` model that has been configured with a `PageListingViewSet` and `EventPageFilterSet`, which has a `CharField` with `choices`, which might be useful as a starting point. There's also `FullFeaturedSnippet` with a `CharField` that has `choices`. Failing that, we can always test a new model. Thanks!

Reply:
@laymonage after some digging I believe that django-filter does not support the dictionary format. This is probably because this feature was only merged about 10 months ago into django. I don't think we should support the dict case if we can't even define a filter with that syntax.

The test makes sense, of course, so I'll add that.

Reply:
Here is the relevant issue in django: https://code.djangoproject.com/ticket/31262